git: throw IOException around submodule update failures

jacob-keller opened the pull request:
Since these commands can fail due to the same reason as git checkout,
(too many concurrent accesses to the git server, for example), we need
to throw an exception caught by the retry loop. Based on inspection, it
appears that the best choice to catch is GitException, since that's what
is thrown by the launchCommand flow that calls the git submodule
command.

jacob-keller commented:
JENKINS-31934

jacob-keller commented:
This change is useful since otherwise issues that are resolved by retry when parent submodule fails to clone are not caught when the same issue occurs to submodules. The primary issue here is when the remote server disconnects with too many concurrent connections due to the number of jobs spawning at once. Retry works to fix this, but submodules are not fixed because they don't throw an IO exception, and the retry handler doesn't kick in.

scottj97 commented:
I need this fix for JENKINS-32481

schniedergers commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -83,16 +83,20 @@ public void onClean(GitSCM scm, GitClient git) throws IOException, InterruptedEx
+                    .ref(build.getEnvironment(listener).expand(reference))
+                    .timeout(timeout)
+                    .execute();
+        } catch (GitException e) {
```
I believe you're missing the closing "}" for the if statement here.

jacob-keller commented:
@MarkEWaite, I revived this change, rebased it, and added a new test case. This should be able to resolve your concerns at https://github.com/jenkinsci/git-plugin/pull/477
